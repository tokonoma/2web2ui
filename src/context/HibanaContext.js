import React, { createContext, useContext } from 'react';
import { connect } from 'react-redux';
import { updateUserUIOptions } from 'src/actions/currentUser';
import { showAlert } from 'src/actions/globalAlert';
import { isUserUiOptionSet } from 'src/helpers/conditions/user';
import { isAccountUiOptionSet } from 'src/helpers/conditions/account';
import { selectCondition } from 'src/selectors/accessConditionState';

const HibanaStateContext = createContext();

function Provider(props) {
  const { isHibanaEnabled: _isHibanaEnabled, children, ...rest } = props;

  // TODO: When OG theme is removed, the env variables will no longer be necessary
  const isHibanaEnabled =
    process.env.REACT_APP_DEFAULT_TO_HIBANA === 'true' ? true : _isHibanaEnabled;

  return (
    <HibanaStateContext.Provider value={{ isHibanaEnabled, ...rest }}>
      {children}
    </HibanaStateContext.Provider>
  );
}

function mapStateToProps(state) {
  return {
    isHibanaEnabled: selectCondition(isUserUiOptionSet('isHibanaEnabled'))(state),
    isBannerVisible: selectCondition(isUserUiOptionSet('isHibanaBannerVisible'))(state),
    hasThemeControls: selectCondition(isAccountUiOptionSet('hibana.hasThemeControls'))(state),
  };
}

const mapDispatchToProps = {
  dismissBanner: () => updateUserUIOptions({ isHibanaBannerVisible: false }),
  setIsHibanaEnabled: bool => {
    if (window.pendo && window.pendo.track) {
      window.pendo.track(`Hibana Toggle - ${Boolean(bool) ? 'On' : 'Off'}`);
    }

    // Always dismiss the banner when the user toggles their theme
    return updateUserUIOptions({ isHibanaEnabled: bool, isHibanaBannerVisible: false });
  },
  showAlert,
};

export function useHibana() {
  const context = useContext(HibanaStateContext);

  if (context === undefined) throw new Error('useHibana must be used within a HibanaProvider');

  return [context];
}

export const HibanaProvider = connect(mapStateToProps, mapDispatchToProps)(Provider);

export function HibanaConsumer({ children }) {
  return <HibanaStateContext.Consumer>{children}</HibanaStateContext.Consumer>;
}
