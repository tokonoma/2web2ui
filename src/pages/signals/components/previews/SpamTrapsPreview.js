import React from 'react';
import { getSpamHits } from 'src/actions/signals';
import { selectSpamHitsDetails } from 'src/selectors/signals';
import { PanelLoading } from 'src/components';
import { PageLink } from 'src/components/links';
import { Panel } from 'src/components/matchbox';
import Callout from 'src/components/callout';
import { useHibana } from 'src/context/HibanaContext';
import { SPAM_TRAP_INFO } from '../../constants/info';
import { spamTrapHitTypesCollection } from '../../constants/spamTrapHitTypes';
import withDetails from '../../containers/withDetails';
import BarChart from '../charts/barchart/BarChart';
import ChartHeader from '../ChartHeader';
import { setSubaccountQuery } from 'src/helpers/subaccounts';
import { roundToPlaces } from 'src/helpers/units';

export function SpamTrapsPreview(props) {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;

  const renderContent = () => {
    const { data, gap, empty, error } = props;

    if (error) {
      return (
        <Callout height="170px" title="Unable to Load Data">
          {error.message}
        </Callout>
      );
    }

    if (empty) {
      return (
        <Callout height="170px" title="No Data Available">
          Insufficient data to populate this chart
        </Callout>
      );
    }

    return (
      <BarChart
        height={170}
        disableHover
        isLink
        margin={{ top: 12, left: 12, right: 0, bottom: 12 }}
        gap={gap}
        timeSeries={data}
        yKeys={spamTrapHitTypesCollection
          .map(({ OGFill, hibanaFill, key }) => {
            const fill = isHibanaEnabled ? hibanaFill : OGFill;

            return {
              fill,
              key: `relative_${key}`,
            };
          })
          .reverse()}
        yAxisProps={{
          tickFormatter: tick => `${roundToPlaces(tick * 100, 2)}%`,
          domain: data.every(({ relative_trap_hits }) => !relative_trap_hits)
            ? [0, 1]
            : ['auto', 'auto'],
        }}
        xAxisProps={{ hide: true }}
      />
    );
  };

  const { facet, facetId, loading, subaccountId } = props;

  if (loading) {
    return <PanelLoading minHeight="170px" />;
  }

  return (
    <PageLink to={`/signals/spam-traps/${facet}/${facetId}${setSubaccountQuery(subaccountId)}`}>
      <Panel>
        <Panel.Section>
          <ChartHeader title="Spam Trap Monitoring" hideLine tooltipContent={SPAM_TRAP_INFO} />
        </Panel.Section>
        <Panel.Section>{renderContent()}</Panel.Section>
      </Panel>
    </PageLink>
  );
}

export default withDetails(SpamTrapsPreview, { getSpamHits }, selectSpamHitsDetails);
