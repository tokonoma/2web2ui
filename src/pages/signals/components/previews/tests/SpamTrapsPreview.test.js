import React from 'react';
import { shallow } from 'enzyme';
import { useHibana } from 'src/context/HibanaContext';
import { SpamTrapsPreview } from '../SpamTrapsPreview';

jest.mock('src/context/HibanaContext');

useHibana.mockImplementation(() => [{ isHibanaEnabled: false }]);

describe('Signals SpamTrapsPreview Component', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      facet: 'ip-pool',
      facetId: 'best-pool',
      loading: false,
      gap: 0.5,
      empty: false,
      data: [1, 2, 3],
      subaccountId: 101,
    };
    wrapper = shallow(<SpamTrapsPreview {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders empty correctly', () => {
    wrapper.setProps({ empty: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders error correctly', () => {
    wrapper.setProps({ error: { message: 'error message' } });
    expect(wrapper).toMatchSnapshot();
  });

  it('gets y axis props with default domain', () => {
    wrapper.setProps({ data: [{ relative_trap_hits: null }] });
    const axisProps = wrapper.find('BarChart').prop('yAxisProps');
    expect(axisProps.tickFormatter(0.2468)).toEqual('24.68%');
    expect(axisProps.domain).toEqual([0, 1]);
  });

  it('gets y axis with domain', () => {
    wrapper.setProps({ data: [{ relative_trap_hits: 0.1 }] });
    const axisProps = wrapper.find('BarChart').prop('yAxisProps');
    expect(axisProps.tickFormatter(0.2468)).toEqual('24.68%');
    expect(axisProps.domain).toEqual(['auto', 'auto']);
  });
});
