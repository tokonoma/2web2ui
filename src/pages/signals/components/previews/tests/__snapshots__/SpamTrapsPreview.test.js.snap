// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals SpamTrapsPreview Component renders correctly 1`] = `
<PageLink
  to="/signals/spam-traps/ip-pool/best-pool?subaccount=101"
>
  <Panel>
    <Panel.Section>
      <ChartHeader
        hideLine={true}
        title="Spam Trap Monitoring"
        tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
      />
    </Panel.Section>
    <Panel.Section>
      <BarChart
        disableHover={true}
        gap={0.5}
        height={170}
        isLink={true}
        margin={
          Object {
            "bottom": 12,
            "left": 12,
            "right": 0,
            "top": 12,
          }
        }
        shouldHighlightSelected={true}
        timeSeries={
          Array [
            1,
            2,
            3,
          ]
        }
        width="99%"
        xAxisProps={
          Object {
            "hide": true,
          }
        }
        xAxisRefLines={Array []}
        xKey="date"
        yAxisProps={
          Object {
            "domain": Array [
              0,
              1,
            ],
            "tickFormatter": [Function],
          }
        }
        yAxisRefLines={Array []}
        yKey="value"
        yKeys={
          Array [
            Object {
              "fill": "#1A838B",
              "key": "relative_trap_hits_typo",
            },
            Object {
              "fill": "#29B9C7",
              "key": "relative_trap_hits_recycled",
            },
            Object {
              "fill": "#50CFDA",
              "key": "relative_trap_hits_parked",
            },
          ]
        }
        yRange={
          Array [
            "auto",
            "auto",
          ]
        }
      />
    </Panel.Section>
  </Panel>
</PageLink>
`;

exports[`Signals SpamTrapsPreview Component renders empty correctly 1`] = `
<PageLink
  to="/signals/spam-traps/ip-pool/best-pool?subaccount=101"
>
  <Panel>
    <Panel.Section>
      <ChartHeader
        hideLine={true}
        title="Spam Trap Monitoring"
        tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
      />
    </Panel.Section>
    <Panel.Section>
      <Callout
        height="170px"
        title="No Data Available"
      >
        Insufficient data to populate this chart
      </Callout>
    </Panel.Section>
  </Panel>
</PageLink>
`;

exports[`Signals SpamTrapsPreview Component renders error correctly 1`] = `
<PageLink
  to="/signals/spam-traps/ip-pool/best-pool?subaccount=101"
>
  <Panel>
    <Panel.Section>
      <ChartHeader
        hideLine={true}
        title="Spam Trap Monitoring"
        tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
      />
    </Panel.Section>
    <Panel.Section>
      <Callout
        height="170px"
        title="Unable to Load Data"
      >
        error message
      </Callout>
    </Panel.Section>
  </Panel>
</PageLink>
`;

exports[`Signals SpamTrapsPreview Component renders loading correctly 1`] = `
<PanelLoading
  accent={false}
  minHeight="170px"
/>
`;
